README.md
<h1>htmlPy</h1>
<h4>HTML5-CSS3-Javascript based GUI library in Python</h4>
<a href="https://travis-ci.org/amol-mandhane/htmlPy" class="badges" target="_blank">
    <img src="https://img.shields.io/travis/amol-mandhane/htmlPy/master.svg">
</a>
<a href="https://pypi.python.org/pypi/htmlPy/" class="badges" target="_blank">
    <img style="max-width:100%;" src="https://img.shields.io/pypi/v/htmlPy.svg">
</a>
<a href="https://pypi.python.org/pypi/htmlPy/" class="badges" target="_blank">
    <img style="max-width:100%;" src="https://img.shields.io/pypi/dm/htmlPy.svg">
</a>
<br><br>

<p>htmlPy is a wrapper around PyQt5's QtWebKit library. It helps with creating beautiful GUIs using <b>HTML5, CSS3 and Javascript</b> for standalone Python applications. It is built on <a href="http://qt.io/" target="_blank">Qt</a> which makes it highly <b>customizable and cross-platform</b>. htmlPy is compatible with both <b>Python2 and Python3</b>. It can be used with any python library or environment like <a href="https://www.djangoproject.com/" target="_blank">django</a>, <a href="http://flask.pocoo.org/" target="_blank">flask</a>, <a href="http://www.scipy.org/" target="_blank">scipy</a>, <a href="http://virtualenv.readthedocs.org/" target="_blank">virtualenv</a> etc. You can use front-end libraries and frameworks like <a href="http://getbootstrap.com/" target="_blank">bootstrap</a>, <a href="http://jquery.com/" target="_blank">jQuery</a>, <a href="http://jqueryui.com/" target="_blank">jQuery UI</a> etc. and create GUIs for your applications in no time.</p>

<h2>Documentation</h2>
<p>The documentation is hosted at <a href="http://htmlpy.readthedocs.org/">http://htmlpy.readthedocs.org/</a>. It contains <b>installation instructions, tutorials, reference guide</b>, compatibility details, and more.</p>

<h2>Example</h2>
<table style="width: 150%; margin-left: -25%;">
    <tr>
        <td>
        <h3>Back-end <br> <small class="typewriter">back_end.py</small></h3>
        <pre>
            <code class="language-python">
import htmlPy


class BackEnd(htmlPy.Object):

    def __init__(self, app):
        super(BackEnd, self).__init__()
        self.app = app

    @htmlPy.Slot()
    def say_hello_world(self):
        self.app.html = u"Hello, world"
                    </code>

                </pre>
        </td>
        <td>
        <h3>GUI <br> <small class="typewriter">main.py</small></h3>
            <pre>
                <code class="language-python">
import htmlPy
from back_end import BackEnd

app = htmlPy.AppGUI(
    title=u"Sample application")
app.maximized = True
app.template_path = "."
app.bind(BackEnd(app))

app.template = ("index.html", {})

if __name__ == "__main__":
app.start()
            </code>
        </pre></td>
        <td>
        <h3>Front-end <br> <small class="typewriter">index.html</small></h3>
        <pre>
            <code class="language-markup highlight">
&lt;html&gt;
  &lt;body&gt;
&lt;a
href="BackEnd.say_hello_world"
data-bind="true"&gt;
  Click to say "Hello, world"
&lt;/a&gt;
  &lt;/body&gt;
&lt;/html&gt;
            </code>
        </pre></td>
    </tr>
</table>

</div>


CHANGES.txt
v2.0.0, 20th Aug 2015 -- Switch from PyQt4 to PySide with improved API
v2.0.1, 20th Aug 2015 -- Add compatibility for Python3
v2.0.2, 21st Aug 2015 -- Add additional files to distutils script
v2.0.3, 22nd Nov 2015 -- Switch to setuptools
v2.0.4  25th May 2017 -- Switch from PySide to PyQt5


htmlPy/app_gui.py
import os

import jinja2

from . import Object, unicode, settings, descriptors
from .base_gui import BaseGUI
from .gui_helper import GUIHelper as Helper


class AppGUI(BaseGUI):
    """ GUI class for creating apps using PyQt5's QtWebkit.

    The class AppGUI can be used to create standalone applications with HTML
    GUI. It uses Jinja2 templating engine for generating HTML which can be
    overridden.

    Note:
        Arguments and Attributes of this class come from the parent class
        :py:class:`htmlPy.BaseGUI`. Please refer to its documentation for more
        details.

    Keyword arguments:
        title (Optional[unicode]): The title of the ``window``. Defaults to
            u"Application".
        width (Optional[int]): Width of the ``window`` in pixels. Defaults to
            800 px. Redundant if ``maximized`` is ``True``.
        height (Optional[int]): Height of the ``window`` in pixels. Defaults to
            600 px. Redundant if ``maximized`` is ``True``.
        x_pos (Optional[int]): The X-coordinate for top-left corner of the
            ``window`` in pixels. Defaults to 10 px. Redundant if
            ``maximized`` is ``True``.
        y_pos (Optional[int]): The Y-coordinate for top-left corner of the
            ``window`` in pixels. Defaults to 10 px. Redundant if
            ``maximized`` is ``True``.
        maximized (Optional[bool]): ``window`` is maximized when set to
            ``True``. Defaults to ``False``.
        plugins (Optional[bool]): Enables plugins like flash when set as
            ``True``. Defaults to ``False``.
        developer_mode (Optional[bool]): Enables developer mode when set as
            ``True``. Defaults to ``False``. The developer mode gives access
            to web inspector and other development tools and enables
            right-click on the webpage.
        allow_overwrite (Optional[bool]): ``PyQt5.QtGui.QApplication`` can be
            instantiated only once. If it is already instantiated, then setting
            ``allow_overwrite`` to ``True`` overwrites the ``QApplication``'s
            window with window of this class instance. If ``False``,
            RuntimeError is raised. If ``QApplication`` is not instantiated,
            this is irrelevent.


    Attributes:
        app (PyQt5.QtGui.QApplication): The singleton Qt application object.
            This can be instantiated only once in the entire process.
        window (PyQt5.QtGui.QMainWindow): The window being displayed in the
            ``app``.
        window (PyQt5.QtWebKit.QWebView): The web view widget which renders
            and displays HTML in the a ``window``.
        html (unicode property): The HTML currently rendered in the
            ``window``. The HTML in ``window`` can be changed by assigning
            the new HTML to this property.
        static_path (str property): The absolute path relative to which the
            ``staticfile`` filter will create links in templating. Changing
            this creates a function dynamically which replaces current
            ``staticfile`` filter in current templating environment.
        template_path (str property): The absolute path relative to which
            jinja2 finds the templates to be rendered. Changing this updates
            the template loader in current templating environment.
        template (tuple(str, dict) property): The current template being
            displayed in ``window``. First element of the tuple is the
            path of the template file relative to ``template_path``. The
            second element of the tuple is the context dictionary in which it
            is being rendered.
        maximized (bool property): A boolean which describes whether the
            ``window`` is maximized or not. Can be set to ``True`` to maximize
            the window and set to ``False`` to restore.
        width (int property): Width of the ``window`` in pixels. Set the value
            of this property in pixels to change the width.
        height (int property): Height of the ``window`` in pixels. Set the
            value of this property in pixels to change the height.
        x_pos (int property): The X-coordinate for top-left corner of the
            ``window`` in pixels. Set the value of this property in pixels to
            move the ``window`` horizontally.
        y_pos (int property): The Y-coordinate for top-left corner of the
            ``window`` in pixels. Set the value of this property in pixels to
            move the ``window`` vertically.
        title (unicode property): The title of the ``window``. Set the value of
            this property to change the title.
        plugins (bool property): A boolean flag which indicates whether plugins
            like flash are enabled or not. Set the value to ``True`` or
            ``False`` as required.
        developer_mode (bool property): A boolean flag which indicated whether
            developer mode is active or not. The developer mode gives access
            to web inspector and other development tools and enables
            right-click on the webpage. Set the value to ``True`` or ``False``
            as required.

    Raises:
        RuntimeError: If ``PyQt5.QtGui.QApplication`` is already instantiated
            and ``allow_overwrite`` is ``False``.

    """

    def __init__(self, *args, **kwargs):
        """ Constructor for :py:class:`htmlPy.WebAppGUI` class """
        super(AppGUI, self).__init__(*args, **kwargs)

        self.__driver_script_location = os.path.abspath(
            os.path.dirname(__file__))

        self.__template = (None, None)
        self.__templating_environment = jinja2.Environment()
        self.static_path = self.__driver_script_location
        self.template_path = self.__driver_script_location

        self.__signal_objects = {Helper.__name__: Helper()}
        self.window.loadFinished.connect(self.__rebind_all)

        script_location = os.path.join(os.path.abspath(
            os.path.dirname(settings.__file__)), "binder.js")

        with open(script_location) as f:
            binder_script = f.read()

        self._javascript_settings = {"binder": binder_script}

    html = descriptors.LiveProperty(
        unicode,
        lambda instance: instance.window.page().mainFrame().toHtml(),
        lambda instance, value: instance.window.setHtml(value))

    static_path = descriptors.CustomAssignmentProperty(
        "static_path", str,
        lambda instance, value: instance._change_static_path(value))

    template_path = descriptors.CustomAssignmentProperty(
        "template_path", str,
        lambda instance, value: instance._change_template_path(value))

    @property
    def template(self):
        """
        tuple(str, dict): The current template being displayed in ``window``.
            First element of the tuple is the path of the template file
            relative to ``template_path``. The second element of the tuple is
            the context dictionary in which it is being rendered.
        """
        return self.__template

    @template.setter
    def template(self, template_tuple):
        """ Display the template represented by ``template_tuple``

        This function renders the template represented by ``template_tuple``
        using jinja2 and sets the result as the HTML of current ``window``.
        """
        if not isinstance(template_tuple, tuple):
            raise TypeError("A tuple is required")
        template_path, context = template_tuple
        if not isinstance(template_path, str):
            raise TypeError("Template path should be a string")
        if not isinstance(context, dict):
            raise TypeError("Context should be a dictionary")

        template = self.__templating_environment.get_template(template_path)
        self.html = template.render(**context)
        self.__template = template_tuple

    def _change_static_path(self, value):
        """ Changes the static_path for jinja2 filters

        NEVER EVER call this function. It is meant for internal use only.
        """
        def staticfile_filter(file_path):
            return "file:///" + os.path.join(value, file_path)

        self.__templating_environment.filters["staticfile"] = staticfile_filter

    def _change_template_path(self, value):
        """ Changes the template_path for jinja2 environment

        NEVER EVER call this function. It is meant for internal use only.
        """
        self.__templating_environment.loader = jinja2.FileSystemLoader(value)

    def bind(self, signal_object, variable_name=None):
        """ Binds an object to be called from GUI javascript.

        This function binds an object to the javascript window of the page. The
        ``signal_object`` should be inherited from :py:class:`htmlPy.Object`.
        The methods that should be callable from javascript should be decorated
        with :py:class:`htmlPy.Slot`. A variable name can be supplied which
        will be the name of the variable in javascript corresponding to that
        object. Otherwise, name of the class of that object will be used as the
        variable name

        Arguments:
            signal_object (htmlPy.Object): The object that has to be bound to
                GUI javascript.

        Keyword Arguments:
            variable_name (str): The name of the javascript variable the object
                should be attached to. Defaults to None. If None,
                ``signal_object.__class__.__name__`` is used.

        Raises:
            TypeError: If ``signal_object`` is not of type
                :py:class:`htmlPy.Object`.
            NameError: If ``variable_name`` is "GUIHelper" or name of the class
                of ``signal_object`` is "GUIHelper"

        """
        if not isinstance(signal_object, Object):
            raise TypeError("The object must be of type htmlPy.Object.")
        name = signal_object.__class__.__name__ if variable_name is None else \
            variable_name
        if name == Helper.__class__.__name__:
            raise NameError("{} is htmlPy reserved name".format(name))
        self.window.page().mainFrame().addToJavaScriptWindowObject(
            name, signal_object)
        self.__signal_objects[name] = signal_object

    def __rebind_all(self):
        """ Rebinds all the objects previously binded.

            Rebinds all the objects stored in the dictionary
            ``self.__signal_objects`` with the dictionary keys as names.
        """
        for name, signal_object in self.__signal_objects.items():
            self.window.page().mainFrame().addToJavaScriptWindowObject(
                name, signal_object)


htmlPy/gui_helper.py
import json

from PyQt5.QtWidgets import QFileDialog

import htmlPy


class GUIHelper(htmlPy.Object):
    """ A class which adds some essential functionalities to GUI javascript.

    An instance of this class is binded to GUI javascript by default. Do not
    use this class manually.

    """

    @htmlPy.Slot(str)
    def log_to_console(self, string):
        """ Prints the string to python console.

        The method is binded to GUI javascript. Thus, the string comes from GUI

        Arguments:
            string (str): The string to be printed.
        """
        print(string)

    @htmlPy.Slot(str, result=str)
    def file_dialog(self, filters="[]"):
        """ Opens a file selection dialog with given extension filter.

        HTML file inputs cannot be directly used with :py:class:`htmlPy.AppGUI`
        . This function, when binded to GUI javascript, gives a method to open
        a file dialog from javascript. With help of ``binder.js``, this task
        is automated with HTML file input.

        Keyword arguments:
            filters (str): A JSON array of javascript objects of type {"title":
                str (Title of the file extension), "extensions": str (space
                separated list of extension wildcards)}. Example ``[{"title":
                "JPEG files", "extensions": "*.jpg *.jpeg"}, {"title":
                "PNG files", "extensions": "*.png"}]``

        """
        extensions = json.loads(filters)
        extensions_filter = ";;".join(map(lambda e: "{} ({})".format(
            e["title"], e["extensions"]), extensions))
        window = QFileDialog()
        return QFileDialog.getOpenFileName(window, "Select file", ".",
                                                 extensions_filter)[0]


htmlPy/web_app_gui.py
from PyQt5.QtCore import QUrl

from . import descriptors, unicode
from .base_gui import BaseGUI


class WebAppGUI(BaseGUI):
    """ GUI class for creating web apps using PyQt5's Qt.

    The class WebAppGUI can be used to create web based applications in a
    QtWebKit based browser running on user side. The server for the web app
    can be remote or local. This can be used for quick desktop deployment of
    existing websites. However, for a standalone application, it is recommended
    to used :py:class:`htmlPy.AppGUI` class.

    Note:
        Arguments and Attributes of this class come from the parent class
        :py:class:`htmlPy.BaseGUI`. Please refer to its documentation for more
        details.

    Keyword arguments:
        title (Optional[unicode]): The title of the ``window``. Defaults to
            u"Application".
        width (Optional[int]): Width of the ``window`` in pixels. Defaults to
            800 px. Redundant if ``maximized`` is ``True``.
        height (Optional[int]): Height of the ``window`` in pixels. Defaults to
            600 px. Redundant if ``maximized`` is ``True``.
        x_pos (Optional[int]): The X-coordinate for top-left corner of the
            ``window`` in pixels. Defaults to 10 px. Redundant if
            ``maximized`` is ``True``.
        y_pos (Optional[int]): The Y-coordinate for top-left corner of the
            ``window`` in pixels. Defaults to 10 px. Redundant if
            ``maximized`` is ``True``.
        maximized (Optional[bool]): ``window`` is maximized when set to
            ``True``. Defaults to ``False``.
        plugins (Optional[bool]): Enables plugins like flash when set as
            ``True``. Defaults to ``False``.
        developer_mode (Optional[bool]): Enables developer mode when set as
            ``True``. Defaults to ``False``. The developer mode gives access
            to web inspector and other development tools and enables
            right-click on the webpage.
        allow_overwrite (Optional[bool]): ``PyQt5.QtGui.QApplication`` can be
            instantiated only once. If it is already instantiated, then setting
            ``allow_overwrite`` to ``True`` overwrites the ``QApplication``'s
            window with window of this class instance. If ``False``,
            RuntimeError is raised. If ``QApplication`` is not instantiated,
            this is irrelevent.


    Attributes:
        app (PyQt5.QtGui.QApplication): The singleton Qt application object.
            This can be instantiated only once in the entire process.
        window (PyQt5.QtGui.QMainWindow): The window being displayed in the
            ``app``.
        window (PyQt5.QtWebKit.QWebView): The web view widget which renders
            and displays HTML in the a ``window``.
        url (unicode property): The URL currently being displayed in
            ``window``. Set the property to a URL unicode string to change the
            URL being displayed.
        html (unicode property): The HTML currently rendered in the
            ``window``. This is a readonly property.
        maximized (bool property): A boolean which describes whether the
            ``window`` is maximized or not. Can be set to ``True`` to maximize
            the window and set to ``False`` to restore.
        width (int property): Width of the ``window`` in pixels. Set the value
            of this property in pixels to change the width.
        height (int property): Height of the ``window`` in pixels. Set the
            value of this property in pixels to change the height.
        x_pos (int property): The X-coordinate for top-left corner of the
            ``window`` in pixels. Set the value of this property in pixels to
            move the ``window`` horizontally.
        y_pos (int property): The Y-coordinate for top-left corner of the
            ``window`` in pixels. Set the value of this property in pixels to
            move the ``window`` vertically.
        title (unicode property): The title of the ``window``. Set the value of
            this property to change the title.
        plugins (bool property): A boolean flag which indicates whether plugins
            like flash are enabled or not. Set the value to ``True`` or
            ``False`` as required.
        developer_mode (bool property): A boolean flag which indicated whether
            developer mode is active or not. The developer mode gives access
            to web inspector and other development tools and enables
            right-click on the webpage. Set the value to ``True`` or ``False``
            as required.

    Raises:
        RuntimeError: If ``PyQt5.QtGui.QApplication`` is already instantiated
            and ``allow_overwrite`` is ``False``.

    """

    def __init__(self, *args, **kwargs):
        """ Constructor for :py:class:`htmlPy.WebAppGUI` class """
        super(WebAppGUI, self).__init__(*args, **kwargs)

    url = descriptors.LiveProperty(
        unicode,
        lambda instance: instance.window.url().toString(),
        lambda instance, link: instance.window.setUrl(QUrl(link)))

    @property
    def html(self):
        """ unicode: The HTML currently rendered in the window.

        This property will return the HTML which is being displayed in the
        ``window``. This is not asynchronous. The URL set with htmlPy will not
        load until the window is in display.
        """

        return self.window.page().mainFrame().toHtml()


htmlPy/base_gui.py
import abc
import sys

from PyQt5.QtCore import QUrl
from PyQt5.QtWebKit import QWebSettings
from PyQt5.QtWebKitWidgets import QWebView
from PyQt5.QtWidgets import QApplication

from . import settings, descriptors, unicode

RIGHT_CLICK_SETTING_KEY = "RIGHT_CLICK"
RIGHT_CLICK_ENABLE = "document.oncontextmenu = null;"
RIGHT_CLICK_DISABLE = "document.oncontextmenu = function(e) { return false; };"
RIGHT_CLICK_INPUTS_ONLY = "document.oncontextmenu = function(e) {" + \
    "return e.target.nodeName == 'INPUT' || e.target.nodeName == 'TEXTAREA'};"

TEXT_SELECTION_SETTING_KEY = "TEXT_SELECTION"
TEXT_SELECTION_ENABLE = "document.onselectstart = null;" + \
    "document.body.style.webkitUserSelect = '';" + \
    "document.body.style.cursor = '';"
TEXT_SELECTION_DISABLE = "document.onselectstart = function(e) " + \
    "{ return false; }; document.body.style.webkitUserSelect = 'none';" + \
    "document.body.style.cursor = 'default';"


class BaseGUI(object):
    """ Abstract GUI class for creating apps using PyQt5's Qt and HTML.

    This class shouldn't be used directly. It serves as a parent to other
    GUI classes. Use :py:class:`htmlPy.AppGUI` and
    :py:class:`htmlPy.WebAppGUI` for developing applications.

    Arguments:
        No args: This is an abstract base class. It must not be instantiated.

    Attributes:
        app (PyQt5.QtGui.QApplication): The singleton Qt application object.
            This can be instantiated only once in the entire process.
        window (PyQt5.QtGui.QMainWindow): The window being displayed in the
            ``app``.
        window (PyQt5.QtWebKit.QWebView): The web view widget which renders
            and displays HTML in the a ``window``.
        maximized (bool property): A boolean which describes whether the
            ``window`` is maximized or not. Can be set to ``True`` to maximize
            the window and set to ``False`` to restore.
        width (int property): Width of the ``window`` in pixels. Set the value
            of this property in pixels to change the width.
        height (int property): Height of the ``window`` in pixels. Set the
            value of this property in pixels to change the height.
        x_pos (int property): The X-coordinate for top-left corner of the
            ``window`` in pixels. Set the value of this property in pixels to
            move the ``window`` horizontally.
        y_pos (int property): The Y-coordinate for top-left corner of the
            ``window`` in pixels. Set the value of this property in pixels to
            move the ``window`` vertically.
        title (unicode property): The title of the ``window``. Set the value of
            this property to change the title.
        plugins (bool property): A boolean flag which indicates whether plugins
            like flash are enabled or not. Set the value to ``True`` or
            ``False`` as required.
        developer_mode (bool property): A boolean flag which indicated whether
            developer mode is active or not. The developer mode gives access
            to web inspector and other development tools and enables
            right-click on the webpage. Set the value to ``True`` or ``False``
            as required.

    """

    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def __init__(self, title=u"Application", width=800, height=600,
                 x_pos=10, y_pos=10, maximized=False,
                 plugins=False, developer_mode=False,
                 allow_overwrite=False):
        """ Abstract constructor for the :py:class:`htmlPy.BaseGUI` class """

        app = QApplication.instance()
        if app is not None and not allow_overwrite:
            raise RuntimeError("Another htmlPy application is already running")
        elif app is not None:
            self.app = app
        else:
            self.app = QApplication(sys.argv)

        self.window = Browser()
        self.web_app = self.window.window()  # backwards compatibility
        # self.window.show()
        self.window.settings().setAttribute(
            QWebSettings.LocalContentCanAccessRemoteUrls, True)

        self._javascript_settings = {}
        self.window.loadFinished.connect(self.__javascript_setting_call)

        self._width = width
        self._height = height
        self._x = x_pos
        self._y = y_pos

        self.title = title
        self.plugins = plugins
        self.developer_mode = developer_mode

        self.maximized = maximized
        self.auto_resize()

    width = descriptors.IntegralGeometricProperty("width")
    height = descriptors.IntegralGeometricProperty("height")
    x_pos = descriptors.IntegralGeometricProperty("x")
    y_pos = descriptors.IntegralGeometricProperty("y")

    maximized = descriptors.LiveProperty(
        bool,
        lambda instance: instance.window.isMaximized(),
        lambda instance, value: instance.window.showMaximized() if value else
        instance.window.showNormal() and instance.auto_resize())

    title = descriptors.LiveProperty(
        unicode,
        lambda instance: instance.window.windowTitle(),
        lambda instance, value: instance.window.setWindowTitle(value))

    plugins = descriptors.LiveProperty(
        bool,
        lambda instance: instance.window.settings().testAttribute(
            QWebSettings.PluginsEnabled),
        lambda instance, value: instance.window.settings().setAttribute(
            QWebSettings.PluginsEnabled, value))

    developer_mode = descriptors.LiveProperty(
        bool,
        lambda instance: instance.window.settings().testAttribute(
            QWebSettings.DeveloperExtrasEnabled),
        lambda instance, value: instance.window.settings().setAttribute(
            QWebSettings.DeveloperExtrasEnabled, value))

    def __javascript_setting_call(self):
        """ Re-evaluate javascript settings

        This function re-evaluates all the javascript settings defined in the
        dictionary ``_javascript_settings``. This should be connected
        to the slot ``window.loadFinished``.
        """
        javascript_string = ";".join(self._javascript_settings.values())
        if len(javascript_string) > 0:
            self.evaluate_javascript(javascript_string)

    def auto_resize(self):
        """ Resizes and relocates the ``window`` to previous state

        If the ``window`` is not maximized, this function resizes it to the
        stored dimensions, moves it to the stored location.
        """
        if not self.maximized:
            self.window.resize(self._width, self._height)
            self.window.move(self._x, self._y)

    def start(self):
        """ Starts the application.

        This is not asynchronous. Starting the application will halt the
        further processes. DO NOT start outside the
        ``if __name__ == "__main__":`` conditional
        """
        self.window.show()
        sys.exit(self.app.exec_())

    def stop(self):
        """ Stops the application. Use only to bind with signals.

        The Qt application does not have to be manually stopped. Also, after
        starting the application is stuck in the execution loop and will not go
        further until it is stopped. Calling this function manually is
        redundant. This function exits only to be binded with QSignals to stop
        the application when that signal is emitted.
        """
        self.app.quit()

    def execute(self):
        """ Executes the application without ending the process on its end.

        DO NOT execute this process directly. Use only when
        :py:meth:`htmlPy.BaseGUI.stop` is connected to some signal.
        """
        self.app.exec_()

    def evaluate_javascript(self, javascript_string):
        """ Evaluates javascript in web page currently displayed

        Arguments:
            javascript_string (str): The string of javascript code that has to
                be evaluated.
        """
        self.window.page().mainFrame().evaluateJavaScript(javascript_string)

    def right_click_setting(self, value):
        """ Javascript based setting for right click on the application.

        This function changes the web page's behaviour on right click. Normal
        behaviour is to open a context menu. Enabling right click exhibits that
        behaviour. Right click is enabled by default. Disabling right click
        suppresses context menu for entire page. Enabling right click for
        only inputs suppresses context menu for all elements excepts inputs and
        textarea, which is the recommended option. The arguments provided
        should be from :py:mod:`htmlPy.settings` module as explained further.

        Arguments:
            value (int): should be either ``htmlPy.settings.ENABLE`` (default)
                or ``htmlPy.settings.DISABLE`` or
                ``htmlPy.settings.INPUTS_ONLY`` (recommended)

        """
        if value == settings.ENABLE:
            self.evaluate_javascript(RIGHT_CLICK_ENABLE)
            self._javascript_settings.pop(RIGHT_CLICK_SETTING_KEY, None)
        elif value == settings.DISABLE:
            self.evaluate_javascript(RIGHT_CLICK_DISABLE)
            self._javascript_settings[RIGHT_CLICK_SETTING_KEY] = \
                RIGHT_CLICK_DISABLE
        elif value == settings.INPUTS_ONLY:
            self.evaluate_javascript(RIGHT_CLICK_INPUTS_ONLY)
            self._javascript_settings[RIGHT_CLICK_SETTING_KEY] = \
                RIGHT_CLICK_INPUTS_ONLY
        else:
            raise ValueError("The argument should be either " +
                             "htmlPy.settings.ENABLE or " +
                             "htmlPy.settings.DISABLE or " +
                             "htmlPy.settings.INPUTS_ONLY")

    def text_selection_setting(self, value):
        """ Javascript based setting for text selection in the application.

        This function changes the web page's behaviour on selection of text.
        Normal behaviour is to highlight the selected text. Enabling text
        selection exhibits that behaviour. Text selection is enabled by
        default. Disabling text selection disallows user to select any text on
        the page except for inputs and disables the I-beam cursor for text
        selection. The arguments provided should be from
        :py:mod:`htmlPy.settings` module as explained further.

        Arguments:
            value (int): should be either ``htmlPy.settings.ENABLE`` (default)
                or ``htmlPy.settings.DISABLE`` or

        """

        if value == settings.ENABLE:
            self.evaluate_javascript(TEXT_SELECTION_ENABLE)
            self._javascript_settings.pop(TEXT_SELECTION_SETTING_KEY, None)
        elif value == settings.DISABLE:
            self.evaluate_javascript(TEXT_SELECTION_DISABLE)
            self._javascript_settings[TEXT_SELECTION_SETTING_KEY] = \
                TEXT_SELECTION_DISABLE
        else:
            raise ValueError("The argument should be either " +
                             "htmlPy.settings.ENABLE or " +
                             "htmlPy.settings.DISABLE")


class Browser(QWebView):
    def __init__(self):
        self.view = QWebView.__init__(self)
        self.setWindowTitle('Loading...')
        self.titleChanged.connect(self.adjustTitle)

    def adjustTitle(self):
        self.setWindowTitle(self.title())

    def load(self, url):
        self.setUrl(QUrl(url))

    def getView(self):
        return self.view
